applications/asset_tracker_v2/src/modules/location_module.c

```diff
@@ -0,0 +1,682 @@
+				     MSEC_PER_SEC;
+
+	if (!IS_ENABLED(CONFIG_APP_REQUEST_GNSS_WAIT_FOR_AGPS)) {
+		return true;
+	} else if (agps_wait_threshold_ms < 0) {
+		/* If CONFIG_APP_REQUEST_GNSS_WAIT_FOR_AGPS_THRESHOLD_SEC is set to -1,
+		 * we need to notify the data module that the application module is awaiting
+		 * A-GPS data in order to request GNSS. If not, GNSS will never be
+		 * requested if the initial A-GPS data request fails.
+		 */
+		if (is_agps_processed()) {
+			return true;
+		}
+
+		SEND_EVENT(app, APP_EVT_AGPS_NEEDED);
+		return false;
+
+	} else if ((agps_wait_threshold_ms < uptime_current_ms) || is_agps_processed()) {
+		return true;
+	}
+
+	return false;
+}
+
+static void search_start(void)
+{
+	int err;
+	struct location_config config;
+	enum location_method methods[2];
+	int methods_count = 0;
+	int methods_index_gnss = -1;
+	int methods_index_cellular = -1;
+
+	if (copy_cfg.no_data.neighbor_cell && copy_cfg.no_data.gnss) {
+		return;
+	}
+
+	if (request_gnss() && !copy_cfg.no_data.gnss) {
+		methods[methods_count] = LOCATION_METHOD_GNSS;
+		methods_index_gnss = methods_count;
+		methods_count++;
+	}
+
+	/* If GNSS shouldn't be requested yet, just request cellular location */
+	if (!copy_cfg.no_data.neighbor_cell) {
+		methods[methods_count] = LOCATION_METHOD_CELLULAR;
+		methods_index_cellular = methods_count;
+		methods_count++;
+	}
+
+	if (methods_count == 0) {
+		LOG_WRN("No location methods selected");
+		return;
+	}
+
+	location_config_defaults_set(&config, methods_count, methods);
+
+	/* The reason for having at least 75 seconds sample timeout when
+	 * requesting GNSS data is that the GNSS module on the nRF9160 modem will always
+	 * search for at least 60 seconds for the first position fix after a reboot. This limit
+	 * is enforced in order to give the modem time to perform scheduled downloads of
+	 * A-GPS data from the GNSS satellites.
+	 *
+	 * However, if A-GPS data has been downloaded via the cloud connection and processed
+	 * before the initial GNSS search, the actual GNSS timeout set by the application is used.
+	 *
+	 * Processing A-GPS before requesting GNSS data is enabled by default,
+	 * and the time that the application will wait for A-GPS data before including GNSS
+	 * in sample requests can be adjusted via the
+	 * CONFIG_APP_REQUEST_GNSS_WAIT_FOR_AGPS_THRESHOLD_SEC Kconfig option. If this option is
+	 * set to -1, the application module will not request GNSS unless A-GPS data has been
+	 * processed.
+	 *
+	 * If A-GPS data has been processed the sample timeout can be ignored as the
+	 * GNSS will most likely time out before the sample timeout expires.
+	 *
+	 * When GNSS is requested, set the sample timeout to (GNSS timeout + 15 seconds)
+	 * to let the GNSS module finish ongoing searches before data is sent to cloud.
+	 */
+	if (methods_index_gnss != -1) {
+		config.methods[methods_index_gnss].gnss.timeout =
+			MAX(copy_cfg.gnss_timeout + 15, 75) * MSEC_PER_SEC;
```

Comment:
This has been copied from the existing ATv2 implementation, that is, we don't use configured GNSS timeout directly but there is this +15 and also maximum of the result and 75.

@simensrostad, do we keep the calculation like this?

@tokangas, @hiltunent, can you check the text above regarding A-GPS etc. so that it's all valid.

Reply:
I understood that ATv2 used GNSS in single fix mode so the sentence about scheduled download is not exactly valid. If GNSS is cold started, data sheet says the fix is obtained in 36 seconds in good conditions, but of course this value varies quite a lot. Anyway, in single fix mode the GNSS will stop as soon as the fix is obtained. 
In location lib when GNSS position is requested the GNSS is briefly started and stopped to see if modem needs A-GPS data. There is no mechanism similar to CONFIG_APP_REQUEST_GNSS_WAIT_FOR_AGPS_THRESHOLD_SEC but instead the location lib monitors the state of RRC connection and only starts GNSS when the A-GPS data is downloaded and RRC state is idle.
Also this sentence does not make sense to me "When GNSS is requested, set the sample timeout to (GNSS timeout + 15 seconds) to let the GNSS module finish ongoing searches before data is sent to cloud." In single fix mode, the GNSS will stop as soon as it is able to produce a fix.
Could it be that ATv2 used the GNSS in periodic mode previously but that has been changed and this comment is not up to date anymore?

Reply:
So my proposal would be to just use whatever timeout the application actually wants and let location lib take care of the rest. For example, if you want timeout of 60 seconds, the location lib will check if A-GPS data is needed, fetch it, and start the timer only after that when the GNSS is actually allowed to start searching the satellites.

Reply:
So my suggestion is that we can keep the download of A-GPS data when ATv2 starts but I guess we could remove `request_gnss()` function and conditions set in that function. Location library handles A-GPS data download if modem requests it.
I'd also suggest we set GNSS timeout configured to ATv2 directly to Location library without any calculations.

Reply:
Sounds reasonable. In general I think the timeout set by the application should be respected as best as possible but I see that this is somewhat of a special case when A-GPS is factored in. Since it typically adds a few seconds+-.
But with some proper documentation it should be fine.

Just note that there are several places in the application that is involved when it comes to A/P-GPS.
main.c(app_module): Only requests GNSS searches if A-GPS has been obtained within a timeout. CONFIG_APP_REQUEST_GNSS_WAIT_FOR_AGPS_THRESHOLD_SEC 
data_module.c: Can trigger requesting of A-GPS data on successful connections. Toggled by CONFIG_DATA_AGPS_REQUEST_ALL_UPON_CONNECTION. To feed modem with A-GPS before search is attempted.
						 Deals with encoding of A/P-GPS requests.
cloud_module.c: Processes incoming A/P-GPS payloads.

It would be nice to place as much as possible in the location module related to A-GPS. Especially processing of incoming payloads and handling of if A-GPS data should be requested or not. (Which is both in main.c and the data module currently) The data module should ideally just encode the payload it receives in specific events.

Reply:
> In general I think the timeout set by the application should be respected as best as possible but I see that this is somewhat of a special case when A-GPS is factored in. Since it typically adds a few seconds+-.

Downloading A-GPS data as such is not very time-consuming normally (of course it might be in poor network conditions) but going idle and into PSM mode after that might take tens of seconds which is why we only want to calculate the time when GNSS has time to run. We are trying to explain this also in Location library API for GNSS timeout and we can add similar text to ATv2 documentation:
"Note that this is not real time as experienced by the user. Since GNSS cannot run while LTE is operating, the running time is counted from the instant when GNSS is allowed to start. Library waits until RRC connection is inactive before starting GNSS. If LTE power saving mode (PSM) is enabled and A-GPS is disabled, library waits until modem enters PSM before starting GNSS, thus maximizing uninterrupted operating window and minimizing power consumption."

> It would be nice to place as much as possible in the location module related to A-GPS. Especially processing of incoming payloads and handling of if A-GPS data should be requested or not. (Which is both in main.c and the data module currently) The data module should ideally just encode the payload it receives in specific events.

I think the only handling we should think of leaving in ATv2 is the one where ATv2 requests A-GPS data when getting connected (flagged with CONFIG_DATA_AGPS_REQUEST_ALL_UPON_CONNECTION). However, is this really necessary? I'd imagine we can have some penalty in the first location request, don't we?

Reply:
For the sake of the implementation CONFIG_DATA_AGPS_REQUEST_ALL_UPON_CONNECTION and its corresponding functionality can be removed. I don't see it as particularly necessary. It was initially added to be able to set low GNSS search time right of the bat.

Reply:
In addition to `CONFIG_DATA_AGPS_REQUEST_ALL_UPON_CONNECTION` related functionality I've removed `APP_REQUEST_GNSS_WAIT_FOR_AGPS` so there is not ATv2 triggered AGPS retrieval. Trigger comes now from location library with `LOCATION_EVT_GNSS_ASSISTANCE_REQUEST` and then ATv2 requests AGPS data from the cloud.
This also means that GNSS method is added into the very first location request that ATv2 sends.

Reply:
GNSS timeout and AGPS related issues resolved now.